### DISCUSSION TITLE
Sensor values for driver in Berry script

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#19020] [General]
Can a driver written in Berry be sure to have all sensor values for sensors defined in template/module definition?

**Comment by barbudor** [comment]
Not sure to understand the question but a Berry driver do not make use of the template
A Berry driver cannot create entries in the GPIO list so you can't use the template or module configure to assign gpios to a Berry driver
This must be done as arguments when creating an instance such as in your autoexec.be

**Reply by sfromis** [reply]
If you use `tasmota.read_sensors()` you get the same sensor data as normally published for sensors supported by Tasmota, meaning all active sensor drivers (including ones coded properly in Berry).

**Comment by msetina** [comment]
My question was in a direction of where in sequence of calls for sensor data to drivers comes new driver. If all the drivers for sensors in hardware are called first, so the latest values are available for new driver.
@sfromis the tasmota.read_sensors() would be my choice, but I was not sure from documentation how are sensor data aquuired from drivers to be sure that second level driver uses latest values from sensors when calculating its derived values.

**Reply by s-hadinger** [reply]
The way data is collected depends on each driver. `read_sensors()` does the exact same process as teleperiod. Some (most) drivers will report the last known value

**Reply by sfromis** [reply]
`tasmota.read_sensors()` is by definition the latest values from the regular scanning loop, about every second, but variable by specific sensor driver. Using `tasmota.read_sensors(true)` forces a poll of all sensor drivers. Beware that if you use `tasmota.read_sensors()` from a Berry driver in the `json_append()` callback, you may run into trouble with your `json_append()` being called again, not great. You could also do the sensor polling in `every_second()` to avoid this.

What you calculate would then be available next time your `json_append()` is called, including at `TelePeriod` for publishing data. Of course, if you do a calculation based on data from other drivers, your values would be based on the prior iteration of those drivers being polled.

**Reply by msetina** [reply]
Thank you for heads up. So it is better for this type of driver to have its own loop like every_second() for calculations and use callback for what they are ,a part of reporting infrastructure.
